Applications/Spire/Source/Ui/TextBox.cpp

```diff
@@ -105,154 +105,289 @@ struct TextBox::TextValidator : QValidator {
-      m_text_color = {};
-      auto& stylist = find_stylist(*parentWidget());
-      auto block = stylist.get_computed_block();
-      for(auto& property : block) {
-        property.visit(
-          [&] (const BorderLeftSize& size) {
-            stylist.evaluate(size, [=] (auto size) {
-              m_margins.setLeft(m_margins.left() + size);
-            });
-          },
-          [&] (const BorderTopSize& size) {
-            stylist.evaluate(size, [=] (auto size) {
-              m_margins.setTop(m_margins.top() + size);
-            });
-          },
-          [&] (const BorderRightSize& size) {
-            stylist.evaluate(size, [=] (auto size) {
-              m_margins.setRight(m_margins.right() + size);
-            });
-          },
-          [&] (const BorderBottomSize& size) {
-            stylist.evaluate(size, [=] (auto size) {
-              m_margins.setBottom(m_margins.bottom() + size);
-            });
-          },
-          [&] (const PaddingLeft& size) {
-            stylist.evaluate(size, [=] (auto size) {
-              m_margins.setLeft(m_margins.left() + size);
-            });
-          },
-          [&] (const PaddingTop& size) {
-            stylist.evaluate(size, [=] (auto size) {
-              m_margins.setTop(m_margins.top() + size);
-            });
-          },
-          [&] (const PaddingRight& size) {
-            stylist.evaluate(size, [=] (auto size) {
-              m_margins.setRight(m_margins.right() + size);
-            });
-          },
-          [&] (const PaddingBottom& size) {
-            stylist.evaluate(size, [=] (auto size) {
-              m_margins.setBottom(m_margins.bottom() + size);
-            });
-          });
+    connection connect_reject_signal(
+        const RejectSignal::slot_type& slot) const {
+      return m_reject_signal.connect(slot);
+    }
+
+  protected:
+    bool eventFilter(QObject* watched, QEvent* event) override {
+      if(event->type() == QEvent::MouseButtonPress) {
+        auto e = static_cast<QMouseEvent*>(event);
+        e->accept();
+        e->setLocalPos(mapFromGlobal(e->globalPos()));
+        mousePressEvent(e);
+      } else if(event->type() == QEvent::KeyPress) {
+        auto e = static_cast<QKeyEvent*>(event);
+        if(e->key() == Qt::Key_Escape) {
+          if(m_submission->get() != m_current->get()) {
+            m_current->set(m_submission->get());
+            return true;
+          }
+        } else if(e->key() == Qt::Key_Enter || e->key() == Qt::Key_Return) {
+          m_has_update = true;
+          on_editing_finished();
+          return true;
+        } else if(!m_is_handling_key_press) {
+          m_is_handling_key_press = true;
+          QLineEdit::keyPressEvent(e);
+          m_is_handling_key_press = false;
+          if(e->isAccepted()) {
+            return true;
+          }
+        }
+      }
+      return QLineEdit::eventFilter(watched, event);
+    }
+    void focusInEvent(QFocusEvent* event) override {
```

Comment:
Vertical whitespace missing.

Reply:
This keeps coming up, so do we always add whitespace after every method definition, regardless of access specifier?

Reply:
No different from ordinary function definitions.